### PR TITLE
adding pytorch to lambdify

### PR DESCRIPTION
**Pull request by clayton-r**
#### References to other Issues or PRs
Fixes #15484

#### Brief description of what is fixed or changed
added mappings to torch printer and fixed merge conflicts from #15484 

#### Other comments
may not work due to pytorch still being too large for TravisCI

#### Release Notes

* printing
  * Allows sympy's lambdify function to operate on pytorch tensors. By example: `f = lambdify(variables, expr, 'torch')` 
  * `torch_code` enables conversion between sympy function and the equivalent pytorch operation.

### CODE REVIEW

**Comment by asmeurer** [comment]
This is good to see. It looks like there are some test failures. 

The release notes entry should be more high level and relevant to end users. See https://github.com/sympy/sympy/wiki/Writing-Release-Notes. They should mention the high level APIs that are added, `lambdify(modules=torch)` and `torch_code`, and not reference implementation details or other PRs.

File: sympy/utilities/lambdify.py [hunk 1]
```diff
@@ -16,16 +16,28 @@
...
+
+# These are the namespaces the lambda functions will use.
+MATH = {}
+MPMATH = {}
+NUMPY = {}
+SCIPY = {}
+TENSORFLOW = {}
+TORCH = {}
+SYMPY = {}
+NUMEXPR = {}
```
Comment (asmeurer) [inline sympy/utilities/lambdify.py:29]: I don't think these lines are necessary. These variables are already defined below.

File: sympy/utilities/lambdify.py [hunk 2]
```diff
@@ -16,16 +16,28 @@
 from sympy.utilities.misc import filldedent
 from sympy.utilities.decorator import doctest_depends_on
 
-__doctest_requires__ = {('lambdify',): ['numpy', 'tensorflow']}
+__doctest_requires__ = {('lambdify',): ['numpy', 'tensorflow', 'torch']}
```
Comment (asmeurer) [inline sympy/utilities/lambdify.py:19]: The doctest examples don't use torch, so this does not need to be added here.

File: sympy/utilities/lambdify.py [hunk 3]
```diff
@@ -165,7 +182,7 @@ def _import(module, reload=False):
 # linecache.
 _lambdify_generated_counter = 1
 
-@doctest_depends_on(modules=('numpy', 'tensorflow', ), python_version=(3,))
+@doctest_depends_on(modules=('numpy', 'tensorflow', 'torch',), python_version=(3,))
```
Comment (asmeurer) [inline sympy/utilities/lambdify.py:185]: Same here, this is not needed.

File: sympy/printing/torch.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        from sympy.codegen.array_utils import CodegenArrayTensorProduct
+        diagonal_indices = list(expr.diagonal_indices)
+        if len(diagonal_indices) > 1:
+            # TODO: this should be handled in sympy.codegen.array_utils,
+            # possibly by creating the possibility of unfolding the
+            # CodegenArrayDiagonal object into nested ones. Same reasoning for
+            # the array contraction.
+            raise NotImplementedError
+        if len(diagonal_indices[0]) != 2:
+            raise NotImplementedError
```
Comment (asmeurer) [inline sympy/printing/torch.py:163]: Can we add messages to these NotImplementedErrors explaining what isn't implemented?

**Comment by asmeurer** [comment]
Someone who's familiar with pytorch will need to review the technical details of what code gets generated here.

File: sympy/printing/pycode.py [hunk 5]
```diff
@@ -185,6 +185,13 @@ def _expand_reduce_binary_op(self, op, args):
                 self._expand_reduce_binary_op(args[Nhalf:]),
             )
 
+    def _get_letter_generator_for_einsum(self):
+        for i in range(97, 123):
+            yield chr(i)
+        for i in range(65, 91):
+            yield chr(i)
```
Comment (eric-wieser) [inline sympy/printing/pycode.py:192]:
```suggestion
        yield from string.ascii_lowercase + string.ascii_uppercase
```

(and `import string`) above

Comment (eric-wieser) [reply sympy/printing/pycode.py:192]: Ah, now I see this code just moved, so fine to leave as is.

File: sympy/printing/torch.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            ", ".join([self._print(i) for i in expr.permutation.args[0]]),
+        )
+
+    def _print_CodegenArrayElementwiseAdd(self, expr):
+        return self._expand_fold_binary_op('torch.add', expr.args)
+
+
+def torch_code(expr):
+    printer = TorchPrinter()
+    return printer.doprint(expr)
```
Comment (eric-wieser) [inline sympy/printing/torch.py:190]:
```suggestion
def torch_code(expr, **settings):
    printer = TorchPrinter()
    return printer.doprint(expr, **settings)
```

File: sympy/printing/torch.py (hunk 4 above)
Comment (Zer0Credibility) [reply sympy/printing/torch.py:163]: I should note that these errors can also be found in printing/tensorflow -- should I add error messages to those as well?

File: sympy/utilities/lambdify.py (hunk 1 above)
Comment (Zer0Credibility) [reply sympy/utilities/lambdify.py]: agreed

File: sympy/utilities/lambdify.py (hunk 2 above)
Comment (Zer0Credibility) [reply sympy/utilities/lambdify.py]: fixed

File: sympy/printing/torch.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+
+class TorchPrinter(AbstractPythonCodePrinter):
+
+    def _get_loop_opening_ending(self, indices):
+        pass
+
+    def _rate_index_position(self, p):
+        pass
+
```
Comment (eric-wieser) [inline sympy/printing/torch.py:15]: Are these used?

File: sympy/printing/torch.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+                self._module_format(op),
+                children[0]
+            )
+        else:
+            return self._expand_fold_binary_op(op, children)
+
+    _print_Expr = _print_Function
+    _print_Application = _print_Function
+    _print_MatrixExpr = _print_Function
+    # TODO: a better class structure would avoid this mess:
```
[the remark this answers is not in the document]
Comment (Zer0Credibility) [reply sympy/printing/torch.py:103]: yes

File: sympy/printing/tests/test_torch.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        skip("Torch not installed")
+
+    M = MatrixSymbol("M", 2, 2)
+    N = MatrixSymbol("N", 2, 2)
+    P = MatrixSymbol("P", 2, 2)
+    Q = MatrixSymbol("Q", 2, 2)
+    # ma = torch.FloatTensor([[1, 2], [3, 4]])
+    # mb = torch.FloatTensor([[1,-2], [-1, 3]])
+    # mc = torch.FloatTensor([[2, 0], [1, 2]])
+    # md = torch.FloatTensor([[1,-1], [4, 7]])
```
[the remark this answers is not in the document]
Comment (Zer0Credibility) [reply sympy/printing/tests/test_torch.py]: removed

File: sympy/printing/torch.py (hunk 7 above)
Comment (Zer0Credibility) [reply sympy/printing/torch.py]: no -- just belong to superclass. Removed

**Comment by ereide** [comment]
Hey. Is this still in development? This feature would be very useful, in particular because of the interoperability with PyTorch with other DNN frameworks and ONNX.

**Comment by Zer0Credibility** [comment]
> Hey. Is this still in development? This feature would be very useful, in particular because of the interoperability with PyTorch with other DNN frameworks and ONNX.

Yes, sorry! Things got busy on my end but I'll go through and fix the CI issues shortly so that it can be merged.

**Comment by czgdp1807** [comment]
@Zer0Credibility Would you like to continue your work on this PR? Please do not close this PR. Thanks for your contributions.

**Comment by Zer0Credibility** [comment]
> @Zer0Credibility Would you like to continue your work on this PR? Please do not close this PR. Thanks for your contributions.

I've opened a new pull request ( #20516 ) that implements the same functionality but with a clean start. That looks like it's going to work.

I've also built a few examples and convenience functions/modules over at https://github.com/Zer0Credibility/symtorch.

**Comment by robinechuca** [comment]
Hello,
I've been waiting a long time for this feature! One of the advantage of most pytorch functions is the 'out' parameter, which allows to calculate inplace. Here's a "TorchPrinter" code to exploit this property: https://framagit.org/robinechuca/cutcutcodec/-/blob/main/cutcutcodec/core/compilation/sympy_to_torch.py
Do you think this feature could be interesting? If so, would you like me to integrate it into this or a future pull request?

**Comment by asmeurer** [comment]
That sounds useful, although I worry that automatically generated in-place code could end up being incorrect due to the difficulty of tracking mutation. Fortunately, unlike C, Python does not have undefined behavior, but it would still require care to make sure everything is correct in every case.

**Comment by asmeurer** [comment]
It seems that work on this pull request has stalled. If you want to see it merged, you may want to start a new pull request based on this one to finish up the work (at a minimum, there are several merge conflicts that need to be resolved). That is unless @clayton-r plans to finish up what was started here.

**Comment by louis-chevallier** [comment]
for some use cases, it is always possible to rely on torch/numpy api similarity as far as math functions are concerned :
dump sympy.solve() result to a file.py, adding 'import torch' and using __import__() to load the function
i've had some success with this trick

**Comment by westurner** [comment]
Who can review, test, verify, or validate these PRs?

Which tests must be added to have a sufficient PR?

- https://github.com/sympy/sympy/pull/20065
- https://github.com/sympy/sympy/pull/20516
  - https://github.com/clayton-r/symtorch

**Comment by oscarbenjamin** [comment]
> Who can review, test, verify, or validate these PRs?

Anyone can do these things. Looking above it seems that some people did do these things.

> Which tests must be added to have a sufficient PR?

I'm not sure. Is it tests that are missing here?

**Comment by oscarbenjamin** [comment]
Anyone wanting this should consider reviving the pull request. Currently it is out of date and has merge conflicts. I suggest doing a rebase to squash this into one commit, rebase over current master and then open a new PR.

The changes here look reasonable although sympy now uses GitHub Actions rather than Travis so that part needs to be updated.

**Comment by oscarbenjamin** [comment]
Closing after https://github.com/sympy/sympy/pull/27760 was merged
